cadc-inventory/src/main/java/org/opencadc/inventory/Entity.java

```diff
@@ -0,0 +1,290 @@
+                + this.getClass().getName(), e);
+        }
+    }
+    
+    private void calcMetaChecksum(Class c, Object o, MessageDigest digest) {
+        // calculation order:
+        // 1. Entity.id for entities
+        // 2. Entity.lastModified?? TBD
+        // 3. state fields in alphabetic order; depth-first recursion
+        // value handling:
+        // Date: truncate time to whole number of seconds and treat as a long
+        // String: UTF-8 encoded bytes
+        // URI: UTF-8 encoded bytes of string representation
+        // float: IEEE754 single (4 bytes)
+        // double: IEEE754 double (8 bytes)
+        // boolean: convert to single byte, false=0, true=1 (1 bytes)
+        // byte: as-is (1 byte)
+        // short: (2 bytes, network byte order == big endian))
+        // integer: (4 bytes, network byte order == big endian)
+        // long: (8 bytes, network byte order == big endian)
+        try {
+            if (o instanceof Entity) {
+                Entity ce = (Entity) o;
+                digest.update(primitiveValueToBytes(ce.id, "Entity.id", digest.getAlgorithm()));
+                if (MCS_DEBUG) {
+                    log.debug("metaChecksum: " + ce.getClass().getSimpleName() + ".id " + ce.id);
+                }
+                // TBD: include lastModified in metaChecksum: CAOM does not but it only supports one-direction
+                // harvesting and always copies lastModified from origin
+            }
+
+            SortedSet<Field> fields = getStateFields(c);
+            for (Field f : fields) {
+                String cf = c.getSimpleName() + "." + f.getName();
+                f.setAccessible(true);
+                Object fo = f.get(o);
+                if (fo != null) {
+                    digest.update(primitiveValueToBytes(fo, cf, digest.getAlgorithm()));
+                } else if (MCS_DEBUG) {
+                    log.debug("skip null: " + cf);
+                }
+            }
+
+        } catch (IllegalAccessException bug) {
+            throw new RuntimeException("Unable to calculate metaChecksum for class " + c.getName(), bug);
+        }
+    }
+    
+    private SortedSet<Field> getStateFields(Class c)
+            throws IllegalAccessException {
+        SortedSet<Field> ret = new TreeSet<>(new FieldComparator());
+        Field[] fields = c.getDeclaredFields();
+        for (Field f : fields) {
+            int m = f.getModifiers();
+            boolean inc = true;
+            inc = inc && !Modifier.isTransient(m);
+            inc = inc && !Modifier.isStatic(m);
+            // no child entities in storage inventory model
+            //inc = inc && !isChildCollection(f); // 0..* relations to other Entity
+            //inc = inc && !isChildEntity(f); // 0..1 relation to other Entity
+            if (inc) {
+                ret.add(f);
+            }
+        }
+        Class sc = c.getSuperclass();
+        while (sc != null && !Entity.class.equals(sc)) {
```

Comment:
Not quite understanding this... isn't the recursion enough?  Why the while loop too?

Reply:
Recursion os for child objects in the tree (not technically used in this model); this traversal is to find all member fields in this object declared here or in super classes. While that's also not used in this model I kept the method as-is because I intend to pull the common parts out of this and caom2 libs to someplace.